### DISCUSSION TITLE
Request for advice on custom wake word model training and evaluating

### DISCUSSION THREAD

**Discussion by sangheonEN** [dscripka/openWakeWord#227] [Q&A]
Hello. I am writing this to explain my current situation and gain an understanding of open wake word learning.

Currently, I am using training_models.ipynb to learn with my dataset.

1. Dataset used:
    - Positive_sample: "hey thomas" / 48 clips
    - Negative_sample: Korean speech audio data / 319 clips

2. Training content: Refer to the attached training_models.ipynb

3. Evaluation Content: I tested it with the attached hey_thomas_female_test_sample.wav file and the prediction value did not increase during the time when hey thomas was spoken, and the prediction value only stayed at the maximum range of 0.008, so it seems that the learning model was not trained properly. See the graph below.

```
scores = oww.predict_clip("/home/openWakeWord/sample_data/hey_thomas_female_test_sample.wav")

plt.figure()
_ = plt.plot([i["hey_thomas2"] for i in scores])
```

![image](https://github.com/user-attachments/assets/09564ebd-00fa-4898-8155-f03ef1efc6f2)

4. Question

**1) Am I doing the training correctly?**
        - An error occurs in the mixed_clips, labels, background_clips = next(mixing_generator) code. (Refer to the attached training_models.ipynb)
        - An error occurs in the openwakeword.data.trim_mmap(output_file) code. (Refer to the attached training_models.ipynb)
        - When I set batch_size = 8 and perform training, the following problem occurs.
        
"---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[40], [line 33](vscode-notebook-cell:?execution_count=40&line=33)
     [31](vscode-notebook-cell:?execution_count=40&line=31) tp = sum(predictions.flatten()[y.flatten() == 1] >= 0.5)
     [32](vscode-notebook-cell:?execution_count=40&line=32) fn = sum(predictions.flatten()[y.flatten() == 1] < 0.5)
---> [33](vscode-notebook-cell:?execution_count=40&line=33) history['recall'].append(float(tp/(tp+fn).detach().numpy()))

AttributeError: 'int' object has no attribute 'detach'"

**2) I would like to get some advice on how to supplement the lack of positive samples.**

Currently, I have about 2 million wav files for negative samples. However, since there are only 50 positive samples, I expect a class imbalance problem to occur. So, the information I found out on my own is as follows. However, I could not find the code for data augmentation.

"https://github.com/dscripka/openWakeWord/blob/main/docs/synthetic_data_generation.md" Here, you mentioned the synthetic data augmentation technique, but I can understand the content, but do I have to develop the code myself? I wonder if there is a code that I can refer to. I don't think I can develop the code myself. I would appreciate it if you could tell me the development method for data augmentation for positive samples.

[openwakeword_advice_data.zip](https://github.com/user-attachments/files/18235274/openwakeword_advice_data.zip)

**Comment by JocelynVelarde** [comment]
Hello! I’ve been also trying to generate my own synthetic audio files from that repo, still I’m facing some different issues, because I’m not very sure on where to use them.

**Reply by sangheonEN** [reply]
Could you elaborate on how you generated the synthetic data?

And to create a wake word model for custom keywords, I'm trying to load the hey_jarvis model and use it as an initialization parameter to do transfer learning using my custom dataset. Have you tried it?

**Reply by JocelynVelarde** [reply]
yea, i’ve been trying to do something pretty similar to your example, i trained my custom model with the word “frida” and now i’ve been trying to get the performance metrics using the positive test and negative test audios that were generated earlier on

**Reply by sangheonEN** [reply]
Then, can you share the information about the learning dataset of the open wake word (the number of positive and negative samples), whether transfer learning was performed, the learning parameters, and the learning code?

And how many synthetic data were generated? Can you tell me which model was used to perform the operation?

**Reply by JocelynVelarde** [reply]
It was nothing personalized, i only downloaded the audio samples that were generated automatically after training the model with the quickstart guide [here](https://colab.research.google.com/drive/1yyFH-fpguX2BTAW8wSQxTrJnJTM-0QAd?usp=sharing&authuser=1#scrollTo=OqdwOgWJxJ1e)

And the audio files can be found in `my_custom_model/model_name`

**Reply by sangheonEN** [reply]
Are you referring to this file automatic_model_training_simple.ipynb?

if "piper-sample-generator/" not in sys.path:
sys.path.append("piper-sample-generator/")

from generate_samples import generate_samples

I think you are using piper-sample-generator to generate voice data with TTS.

If you simply use this, it doesn't seem to perform well.

So I recorded "hey thomas" myself as custom data, and acquired data from a platform that opened a Korean TTS model. However, since there were only 50, the performance was not good even after training.

So what I'm going to do is to generate 3,000 hey thomas voices using NVIDIA WAVEGLOW and VITS, add them to the positive sample, and apply data augmentation techniques. And I'm going to import the hey jarvis model parameters and do transfer learning.

I think you should look at the code in openwakeword/openwakeword/train.py .

Does the automatic_model_training_simple.ipynb code include what I'm talking about?

# Generate clips
!{sys.executable} openwakeword/openwakeword/train.py --training_config my_model.yaml --generate_clips

# Step 2: Augment the generated clips

!{sys.executable} openwakeword/openwakeword/train.py --training_config my_model.yaml --augment_clips

# Step 3: Train model

!{sys.executable} openwakeword/openwakeword/train.py --training_config my_model.yaml --train_model

I'm curious about how Generate clips works in the above task.

Looking for it in the train.py code,

 args = parser.parse_args()
 config = yaml.load(open(args.training_config, 'r').read(), yaml.Loader)

 # imports Piper for synthetic sample generation
 sys.path.insert(0, os.path.abspath(config["piper_sample_generator_path"]))
 from generate_samples import generate_samples

...

 if args.generate_clips is True:
 # Generate positive clips for training
 logging.info("#"*50 + "\nGenerating positive clips for training\n" + "#"*50)
 if not os.path.exists(positive_train_output_dir):
 os.mkdir(positive_train_output_dir)
 n_current_samples = len(os.listdir(positive_train_output_dir))
 if n_current_samples <= 0.95*config["n_samples"]:
 generate_samples(
 text=config["target_phrase"], max_samples=config["n_samples"]-n_current_samples,
 batch_size=config["tts_batch_size"],
 noise_scales=[0.98], noise_scale_ws=[0.98], length_scales=[0.75, 1.0, 1.25],
 output_dir=positive_train_output_dir, auto_reduce_batch_size=True,
 file_names=[uuid.uuid4().hex + ".wav" for i in range(config["n_samples"])]
 )
torch.cuda.empty_cache()
else:
logging.warning(f"Skipping generation of positive clips for training, as ~{config['n_samples']} already exist")

It is declared like this. That is, it seems to create a piper_sample_generator and utilize the piper tts model. It seems that we need to check whether appropriate samples are generated.

Have you heard the generated clip samples yourself? Can you share the file?

**Reply by JocelynVelarde** [reply]
ohhhhh not cool, yea so i was not aware it did not perform well. I was about to use the synthetic audios to get the performance metrics for my model. So you would rather recommend doing that ourselves? Also I tried using the instructions on this [repo](https://github.com/dscripka/synthetic_speech_dataset_generation) but i just kept getting errors.

**Reply by JocelynVelarde** [reply]
yea ofc i can share the positive files on .zip for you to hear.
[positive_test.zip](https://github.com/user-attachments/files/18285932/positive_test.zip)

**Reply by sangheonEN** [reply]
What error occurred in which process?

First, I set up the development environment of https://github.com/dscripka/synthetic_speech_dataset_generation and generated 5000 positive samples. The next step I will do is to use the dataset to learn according to the code written here, automatic_model_training.ipynb.

**Reply by JocelynVelarde** [reply]
For using the automatic model training it is not mandatory for you to upload your custom files it automatically generates them when you run the code from the google colab.

So the error I was getting is the one I mentioned on the repo 😂 [here](https://github.com/dscripka/synthetic_speech_dataset_generation/issues/1), but as I mentioned I will try it again clean tomorrow.

**Reply by sangheonEN** [reply]
I ran the training by generating 55,000 hey thomas positive samples with the tts(VITS) model. I used 23,000 data from any Korean dataset for the negative samples.

I was able to confirm that the training was done well, as shown in the plot.

![image](https://github.com/user-attachments/assets/ee34ca39-a929-4ea1-9a64-4801c6ead92f)

And when I evaluated it using the test data hey_thomas_female_test_sample.wav file, I was able to confirm that the two triggers were performed well and the prediction value was close to 1, as shown in the plot below.

![image](https://github.com/user-attachments/assets/952acdc4-ce49-4bf7-b71c-b93670c569ed)

However, when I evaluated the hey_thomas3.onnx model learned above on realtimestt (https://github.com/KoljaB/RealtimeSTT), it was not properly activated.

The prediction value stayed at the level of 0.00002~9, and I confirmed that it was activated for voices other than the hey thomas keyword voice. What could be the problem? Is there a problem with the inference because the chunk size of the training data is different from the chunk size of realtimestt?

The model seems to have trained normally, but when I try to actually use it, it doesn't work.

**Reply by sangheonEN** [reply]
I checked and the device sample rate of the microphone I tested is 16000, which is the same as the training data. But why isn't the performance improving?

**Reply by JocelynVelarde** [reply]
hello! so i’m not sure what to say in here, because i’m in the same scenario. I already trained 3 different models and have been trying to obtain the metrics to measure it like you just did but I just can’t seem to make the script for performance_metrics work, i just keep getting errors. So i’m still not sure if my models work efficiently, i’ve tested them with microphone and it seems to be working well, still i want some graphs like the ones you show.

**Reply by sangheonEN** [reply]
Are you saying that when you tested with a microphone, it worked well?

On the contrary, when I tested with a microphone, it didn't work well, but you said that the model trained by following this colab code https://github.com/dscripka/openWakeWord/blob/main/notebooks/automatic_model_training.ipynb worked well? Could you share the code for testing with a microphone?

And do you know if it is possible to generate these models embedding_model.onnx and melspectrogram.onnx using my custom data, as I wrote here: https://github.com/dscripka/openWakeWord/discussions/230?

**Reply by JocelynVelarde** [reply]
yes it worked fine, [this](https://github.com/RoBorregos/home-hri/tree/df9df4b2f14a6fa5c80e0a3ef9796c26b27dc0b5/speech/OpenWakeWord/models) is the model i trained, you can also test it on that script within the repo. As for the model notebook let me share [it](https://colab.research.google.com/drive/1B-5l7ZG9d7C_Z6XhEZUqrlTzVnsJi76Y?usp=sharing). 

I only did two updates to this script. Modifying the `openwakeword/examples/custom_model.yml` file with the wake word I want, on that case the word is “no”. And then just typing “no” on the `config["target_phrase"] = ["no”]`

Have you tried this before?

**Reply by JocelynVelarde** [reply]
As for the melspectrogram.onnx I’m pretty sure you can create your own and just replace it on the downloads at the top of the script. Haven’t tried that tho^^

**Reply by sangheonEN** [reply]
[automatic_model_training_simple.zip](https://github.com/user-attachments/files/18342161/automatic_model_training_simple.zip)
Yes, I also followed this process https://github.com/dscripka/openWakeWord/blob/main/notebooks/automatic_model_training.ipynb and set config["target_phrase"] = ["hey thomas"] to create a hey thomas model. However, the hey thomas voice data that I spoke through the microphone was input to the model and the output prediction value was in the range of 0.001~0.002, so normal learning did not occur. So what I tried is exactly what I explained above.

I created 55,000 samples of Hey thomas through VITS and PIPER TTS and used them as positive samples. Also, I used 23,000 data from any Korean dataset for the negative samples.

Also, the code used is https://github.com/dscripka/openWakeWord/blob/main/notebooks/training_models.ipynb

I used the code.

However, the learning seems to be good, but when I say hey thomas through the microphone, it doesn't work properly, and The prediction value stayed at the level of 0.00002~9.

Should I implement the code that integrates this process with https://github.com/dscripka/openWakeWord/blob/main/notebooks/automatic_model_training.ipynb and this process https://github.com/dscripka/openWakeWord/blob/main/notebooks/training_models.ipynb myself..? I don't know why the performance is not good.

**Reply by sangheonEN** [reply]
When training the model, the "hey thomas" positive sample (class0) is a TTS model generated based on English pronunciation, and the negative sample (class1) is an unspecified number of voice data without "hey thomas". It was trained using about 22,000 voice data provided here (https://www.aihub.or.kr/aihubdata/data/view.do?dataSetSn=109). The difference in pronunciation of positive sample/negative sample will be divided into English/Korean. Will this have an effect?
